luminoso_api/v5_client.py

```diff
@@ -0,0 +1,372 @@
+        anything on the server.
+        """
+        params = jsonify_parameters(params)
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('get', url, params=params)
+
+    def post(self, path='', **params):
+        """
+        Make a POST request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the POST.
+
+        POST requests are requests that cause a change on the server,
+        especially those that ask to create and return an object of some kind.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('post', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def put(self, path='', **params):
+        """
+        Make a PUT request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the PUT.
+
+        PUT requests are usually requests to *update* the object represented by
+        this URL. Unlike POST requests, PUT requests can be safely duplicated.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('put', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def patch(self, path='', **params):
+        """
+        Make a PATCH request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to form values, sent in the body
+        of the PATCH.
+
+        PATCH requests are usually requests to make *small fixes* to the
+        object represented by this URL.
+        """
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('patch', url, data=json.dumps(params),
+                                  headers={'Content-Type': 'application/json'})
+
+    def delete(self, path='', **params):
+        """
+        Make a DELETE request to the given path, and return the JSON-decoded
+        result.
+
+        Keyword parameters will be converted to URL parameters.
+
+        DELETE requests ask to delete the object represented by this URL.
+        """
+        params = jsonify_parameters(params)
+        url = ensure_trailing_slash(self.url + path.lstrip('/'))
+        return self._json_request('delete', url, params=params)
+
+    # Useful abstractions
+    def change_path(self, path):
+        """
+        Change the client's path. Paths with leading slashes are appended to
+        the root url, otherwise paths are set relative to the current path.
+        Returns a copy of the client to avoid breaking old code.
+        """
+        if path.startswith('/'):
+            self.url = ensure_trailing_slash(self.root_url + path)
```

Comment:
Now that I'm looking back at this, I remember that conversation with Kevin a few weeks back that made me think it might be worth walking back the mutate-on-change_path decision. I should try to get a definitive answer on that before merging this.

Reply:
Kevin did indeed seem to prefer the old style, so I've put in (easily-revertible) commits both here and in the tests in lumi_stack_tools to change to the old behavior (and to change the name, since I think part of the motivation was that `change_path` looked like it was changing your object, not generating a new one).